pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java

```diff
@@ -180,27 +199,126 @@ public List<String> getSegmentMetadataFromServer(String tableNameWithType,
     return segmentsMetadata;
   }
 
+  /**
+   * This method is called when the API request is to fetch validDocId metadata for a list segments of the given table.
+   * This method will pick a server that hosts the target segment and fetch the segment metadata result.
+   *
+   * @return segment metadata as a JSON string
+   */
+  public List<ValidDocIdMetadataInfo> getValidDocIdMetadataFromServer(String tableNameWithType,
+      Map<String, List<String>> serverToSegmentsMap, BiMap<String, String> serverToEndpoints,
+      @Nullable List<String> segmentNames, int timeoutMs) {
+    List<Pair<String, String>> serverURLsAndBodies = new ArrayList<>();
+    for (Map.Entry<String, List<String>> serverToSegments : serverToSegmentsMap.entrySet()) {
+      List<String> segmentsForServer = serverToSegments.getValue();
+      List<String> segmentsToQuery = new ArrayList<>();
+      for (String segment : segmentsForServer) {
+        if (segmentNames == null) {
+          // If segmentNames is null, query all segments
+          segmentsToQuery.add(segment);
+        } else if (segmentNames.contains(segment)) {
```

Comment:
Consider using set instead of list for segments (when there are large number of segments) ?

Reply:
changed to set